File: packages/typescript/service/src/storage-redaction.ts
/**
 * Storage-side redaction — strips parameters from low-value audit entries before persistence.
 *
 * The persisted redacted entry is the canonical hashed form for checkpointing.
 * This is independent of response-boundary redaction (disclosure level).
 */

const LOW_VALUE_CLASSES = new Set([
  "low_risk_success",
  "malformed_or_spam",
  "repeated_low_value_denial",
]);

export function storageRedactEntry(
  entry: Record<string, unknown>,
): Record<string, unknown> {
  const result = { ...entry };
  const eventClass = result.event_class as string | undefined;

  if (eventClass != null && LOW_VALUE_CLASSES.has(eventClass)) {
    result.parameters = null;
    result.storage_redacted = true;
  } else {
    result.storage_redacted = false;
  }

  return result;
}

File: packages/go/service/storage_redaction.go
package service

// StorageRedactEntry strips parameters from low-value audit entries
// before persistence. Implements SPEC §6.10.
// Returns a shallow copy — does not mutate the input.
func StorageRedactEntry(entry map[string]any) map[string]any {
	result := make(map[string]any, len(entry))
	for k, v := range entry {
		result[k] = v
	}

	ec, _ := result["event_class"].(string)
	if isLowValueClass(ec) {
		result["parameters"] = nil
		result["storage_redacted"] = true
	} else {
		result["storage_redacted"] = false
	}
	return result
}

func isLowValueClass(ec string) bool {
	return ec == "low_risk_success" || ec == "malformed_or_spam" || ec == "repeated_low_value_denial"
}
